Sources/Apollo/ApolloStore.swift

```diff
@@ -61,25 +57,20 @@ public final class ApolloStore {
   /// - Returns: A promise which fulfills when the Cache is cleared.
   public func clearCache(callbackQueue: DispatchQueue = .main, completion: ((Result<Void, Error>) -> Void)? = nil) {
     queue.async(flags: .barrier) {
-      self.cacheLock.withWriteLock {
-          self.cache.clearPromise()
-        }.andThen {
-          DispatchQueue.apollo.returnResultAsyncIfNeeded(on: callbackQueue,
-                                                         action: completion,
-                                                         result: .success(()))
-      }
+      let result = Result { try self.cache.clear() }
+      DispatchQueue.apollo.returnResultAsyncIfNeeded(on: callbackQueue,
+                                                     action: completion,
+                                                     result: result)
     }
   }
 
-  func publish(records: RecordSet, identifier: UUID? = nil) -> Promise<Void> {
-    return Promise<Void> { fulfill, reject in
-      queue.async(flags: .barrier) {
-        self.cacheLock.withWriteLock {
-          self.cache.mergePromise(records: records)
-        }.andThen { changedKeys in
-          self.didChangeKeys(changedKeys, identifier: identifier)
-          fulfill(())
-        }.wait()
+  func publish(records: RecordSet, identifier: UUID? = nil) {
+    queue.async(flags: .barrier) {
+      do {
+        let changedKeys = try self.cache.merge(records: records)
+        self.didChangeKeys(changedKeys, identifier: identifier)
+      } catch {
+        assertionFailure(String(describing: error))
```

Comment:
We could just allow an optional completion handler. The reason I left it like this was that it isn't clear what callers (actually, that's just `LegacyCacheWriteInterceptor` right now) would do with a failure other than log it. You probably don't want to fail the the whole chain for example. 

The `assertionFailure` should crash the whole app (leaving it in a debuggable state in Xcode). But it will only do that on debug builds, not release builds. If we had some kind of logging set up maybe we could log an error instead.

Reply:
Maybe `assertionFailure` if there's no completion handler, otherwise pass the error to the completion handler?

Reply:
Yep, good idea. I ended up putting the assert in `returnResultAsyncIfNeeded` instead, because avoiding swallowing errors seems like useful behavior for all asynchronous operations that take an optional completion handler (like `clearCache`, `withinReadTransaction`, `withinReadWriteTransaction`). (For comparison, Node.js logs a warning to the console for unhandled promise rejections, and in web environments there is [a global event](https://developer.mozilla.org/en-US/docs/Web/API/Window/unhandledrejection_event) you can register for.)